move go-libp2p-webtransport to p2p/transport/webtransport

marten-seemann opened the pull request:
Part of #1717.

MarcoPolo reviewed (commented):
Some small comments/questions.

Nothing blocking if we agree we can change the multiaddr `quic/webtransport` part later, but if we are committing to this right now I think we should take a bit more. Not saying we need to change it later, just that we agree we _could_ change it.

MarcoPolo commented:
I left a comment instead of an approval. I think another blocker is the potential stalling of the http server. If that's addressed as well as the other comment (`quic/webtransport`) then I think this is good and the rest of the changes can be done in separate PRs.

marten-seemann commented on p2p/transport/webtransport/listener.go:
```diff
@@ -0,0 +1,247 @@
+	ln.ctx, ln.ctxCancel = context.WithCancel(context.Background())
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("Hello, world!"))
```
Yeah, this is a leftover from debugging. Removed it in https://github.com/marten-seemann/go-libp2p-webtransport/commit/9354bc90c054072d418a8adaef72f113ac8ea352.

marten-seemann commented on p2p/transport/webtransport/cert_manager.go:
```diff
@@ -0,0 +1,181 @@
+type certManager struct {
+	clock     clock.Clock
+	ctx       context.Context
+	ctxCancel context.CancelFunc
```
Hmm, this is also the pattern we've been using all across go-libp2p, in every service that spawns Go routines. It's super convenient since you can call the `CancelFunc` multiple times (which you can't do if you do the naive implementation with a `closeChan chan struct{}`, as you can only close a channel once).

MarcoPolo replied:
Yeah I know we do it a lot. Just worth pointing out that I don’t think it’s idiomatic go, and maybe we should stop adding new code like this. 

Not a blocking change of course

marten-seemann commented on p2p/transport/webtransport/multiaddr.go:
```diff
@@ -0,0 +1,68 @@
+	"github.com/multiformats/go-multihash"
+)
+
+var webtransportMA = ma.StringCast("/quic/webtransport")
```
Yes, the IETF is working on a HTTP/2 fallback for WebTransport: https://www.ietf.org/id/draft-ietf-webtrans-http2-03.html. I have no intention of implementing it at this point, but it's there.

MarcoPolo replied:
Got it, then this makes sense to me. Thanks for explaining

marten-seemann commented on p2p/transport/webtransport/transport.go:
```diff
@@ -0,0 +1,267 @@
+	}
+
+	if len(certHashes) > 0 {
+		tlsConf.InsecureSkipVerify = true // this is not insecure. We verify the certificate ourselves.
```
Done in https://github.com/marten-seemann/go-libp2p-webtransport/commit/734b51f23caef06f52c286f695a34ccdf62791d7.

marten-seemann commented on p2p/transport/webtransport/cert_manager.go:
```diff
@@ -0,0 +1,181 @@
+	return nil
+}
+
+func (m *certManager) addrComponentForCert(hash []byte) (ma.Multiaddr, error) {
```
Good catch. Done in https://github.com/marten-seemann/go-libp2p-webtransport/commit/a883b71ed581c566d12cf31b10eb95a2bbcbf229.

marten-seemann commented on p2p/transport/webtransport/conn.go:
```diff
@@ -0,0 +1,65 @@
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type connSecurityMultiaddrs interface {
```
Honestly, I don't remember why I introduced it. I'd blame it on some refactoring ;)

I removed it in https://github.com/marten-seemann/go-libp2p-webtransport/commit/7aa5c3b6130bb8fa9157b1b97a755551b36a06ea.

marten-seemann commented on p2p/transport/webtransport/transport.go:
```diff
@@ -0,0 +1,267 @@
+	if t.gater != nil && !t.gater.InterceptSecured(network.DirOutbound, p, sconn) {
+		// TODO: can we close with a specific error here?
+		sess.Close()
+		return nil, fmt.Errorf("secured connection gated")
```
Good catch! Fixed in https://github.com/marten-seemann/go-libp2p-webtransport/commit/3cc4782bd9b2a59d71cc246054df145199c42ba0.

marten-seemann commented on p2p/transport/webtransport/crypto.go:
```diff
@@ -0,0 +1,108 @@
+	if err != nil {
+		return err
+	}
+	// TODO: is this the best (and complete?) way to identify RSA certificates?
```
I'm not sure how to check for this. We only have access to the parsed certificate, where the signature algorithm is encoded as a [x509.SignatureAlgorithm](https://pkg.go.dev/crypto/x509#SignatureAlgorithm). I'd assume that those other algorithms would show up as `x509.UnknownSignatureAlgorithm`. Maybe it's sufficient to reject that one?

marten-seemann commented on p2p/transport/webtransport/transport.go:
```diff
@@ -0,0 +1,267 @@
+		return nil, err
+	}
+
+	// Now run a Noise handshake (using early data) and verify the cert hash.
```
Updated the comment in https://github.com/marten-seemann/go-libp2p-webtransport/commit/462ca9902b62ddcb5c4fd58d4dbc9f72c20fca37.

marten-seemann commented on p2p/transport/webtransport/listener.go:
```diff
@@ -0,0 +1,247 @@
+	}
+
+	// TODO: think about what happens when this channel fills up
+	l.queue <- newConn(l.transport, sess, sconn, connScope)
```
Fixed in https://github.com/marten-seemann/go-libp2p-webtransport/pull/24.

marten-seemann commented on p2p/transport/webtransport/transport.go:
```diff
@@ -0,0 +1,267 @@
+		return true
+	})
+	if numHashes == 0 {
+		return false
```
Good catch!

We even have a test that asserts that we can dial a node with a static TLS certificate. But this test only used the `Dial` method without checking the `CanDial` result first. I fixed the behavior and the test case in https://github.com/marten-seemann/go-libp2p-webtransport/pull/25.

marten-seemann commented:
@MarcoPolo Thank you for the comprehensive review! I've addressed all your comments and recreated this PR.

> Nothing blocking if we agree we can change the multiaddr quic/webtransport part later, but if we are committing to this right now I think we should take a bit more. Not saying we need to change it later, just that we agree we could change it.

See my comment here: https://github.com/libp2p/go-libp2p/pull/1737#discussion_r964465572.
